http3/conn.go

```diff
@@ -0,0 +1,530 @@
+		return nil, ctx.Err()
+	case <-conn.session.Context().Done():
+		return nil, conn.session.Context().Err()
+	}
+}
+
+func (conn *connection) OpenRequestStream(ctx context.Context) (RequestStream, error) {
+	if conn.session.Perspective() != quic.PerspectiveClient {
+		return nil, errors.New("client method called on server connection")
+	}
+	conn.incomingUniStreamsOnce.Do(func() {
+		go conn.handleIncomingUniStreams()
+	})
+	str, err := conn.session.OpenStreamSync(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return newRequestStream(conn, str, nil), nil
+}
+
+func (conn *connection) handleIncomingStreams() {
+	var wg sync.WaitGroup
+	for {
+		str, err := conn.session.AcceptStream(context.Background())
+		if err != nil {
+			conn.incomingStreamsErr = err
+			// TODO: log the error
+			break
+		}
+		wg.Add(1)
+		go func(str quic.Stream) {
+			conn.handleIncomingStream(str)
+			wg.Done()
+		}(str)
+	}
+	wg.Wait()
+	close(conn.incomingRequestStreams)
+}
+
+func (conn *connection) handleIncomingStream(str quic.Stream) {
+	r := quicvarint.NewReader(str)
+
+	t, err := quicvarint.Read(r)
+	if err != nil {
+		str.CancelWrite(quic.StreamErrorCode(errorRequestIncomplete))
+		return
+	}
+
+	switch FrameType(t) { //nolint:exhaustive
+	case FrameTypeHeaders:
+		b := &bytes.Buffer{}
+		quicvarint.Write(b, t)
+		conn.incomingRequestStreams <- newRequestStream(conn, str, b.Bytes())
+	case FrameTypeWebTransportStream:
+		if !conn.negotiatedWebTransport() {
+			// TODO: log error
+			// TODO: should this close the connection or the stream?
+			// https://github.com/ietf-wg-webtrans/draft-ietf-webtrans-http3/pull/56
+			str.CancelWrite(quic.StreamErrorCode(errorSettingsError))
+			return
+		}
+		id, err := quicvarint.Read(r)
+		if err != nil {
+			str.CancelWrite(quic.StreamErrorCode(errorFrameError))
+			return
+		}
+		select {
+		case conn.incomingStreamsChan(SessionID(id)) <- str:
+		default:
+			str.CancelWrite(quic.StreamErrorCode(errorWebTransportBufferedStreamRejected))
+			return
+		}
+	default:
+		conn.session.CloseWithError(quic.ApplicationErrorCode(errorFrameUnexpected), "expected first frame to be a HEADERS frame")
+		return
+	}
+}
+
+func (conn *connection) handleIncomingUniStreams() {
+	for {
+		str, err := conn.session.AcceptUniStream(context.Background())
+		if err != nil {
+			// TODO: log the error
+			return
+		}
+		// FIXME: This could lead to resource exhaustion.
+		// Chrome sends 2 unidirectional streams before opening the first WebTransport uni stream.
+		// The streams are open, but zero data is sent on them, which blocks reads below.
+		go conn.handleIncomingUniStream(str)
+	}
+}
+
+func (conn *connection) handleIncomingUniStream(str quic.ReceiveStream) {
+	r := quicvarint.NewReader(str)
+	t, err := quicvarint.Read(r)
+	if err != nil {
+		str.CancelRead(quic.StreamErrorCode(errorGeneralProtocolError))
+		return
+	}
+	streamType := StreamType(t)
+
+	// Store control, QPACK, and push streams on conn
+	if streamType < 4 {
```

Comment:
Where does this magic number come from?

Reply:
4 standard H3 streams: control, qpack, and push (0x0-0x3).